### ISSUE TITLE
Supported C++ features document is out of date

### ISSUE THREAD

**Issue by ThanHenderson** [eclipse/omr#7359]
The document here: https://github.com/eclipse/omr/blob/master/doc/SupportedC%2B%2BFeatures.md
outlines the minimum-supported `C++` compiler versions and thus the available `C++` features. However, this document appears to be out of date.

e.g. an excerpt from the OpenJ9 building instructions states:

> ### gcc
> 
> The minimum accepted version of gcc is 5.0. Older versions will generate a warning
> by `configure` and are unlikely to work.
> 
> The JDK is currently known to be able to compile with at least version 11.2 of
> gcc.
> 
> In general, any version between these two should be usable.

but the document talks about 4.4.7 being the minimum supported GCC version, suggesting there is a divergence between the documented supported features and the actual.  e.g. an `rg nullptr` reveals `nullptr` being used a few times in the codebase, but this list suggests that it is unsupported.

We should audit and update this document to get a better idea of which features we can use.

**Comment by ThanHenderson** [comment]
JDK8: https://github.com/ibmruntimes/openj9-openjdk-jdk8/blob/c1470c8cffd73bfdd07ff63c25ff71da14cbb54f/common/autoconf/flags.m4#L528 (at least minimum of 6 for GCC)

JDK11: https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/839808e0d9f5c314f56d26ab57a0b0ac618c0e22/make/autoconf/toolchain.m4#L54-L59

JDK21: https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/cbd5117c34619bfb4b5e6cf11815a118df44bc9a/make/autoconf/toolchain.m4#L52-L56

From the above:

Potentially other relevant info:
https://github.com/eclipse-openj9/openj9/blob/master/doc/release-notes/0.43/0.43.md

**Comment by ThanHenderson** [comment]
From above:

JDK8: only seems to specify minimum for gcc/g++ at v6.0

JDK11:
```
# Minimum supported versions, empty means unspecified
TOOLCHAIN_MINIMUM_VERSION_clang="3.2"
TOOLCHAIN_MINIMUM_VERSION_gcc="4.8"
TOOLCHAIN_MINIMUM_VERSION_microsoft="16.00.30319.01" # VS2010
TOOLCHAIN_MINIMUM_VERSION_solstudio="5.13"
TOOLCHAIN_MINIMUM_VERSION_xlc=""
```

JDK17:
```
# Minimum supported versions, empty means unspecified
TOOLCHAIN_MINIMUM_VERSION_clang="3.5"
TOOLCHAIN_MINIMUM_VERSION_gcc="5.0"
TOOLCHAIN_MINIMUM_VERSION_microsoft="19.10.0.0" # VS2017
TOOLCHAIN_MINIMUM_VERSION_xlc=""
```
JDK21:
```
# Minimum supported versions, empty means unspecified
TOOLCHAIN_MINIMUM_VERSION_clang="3.5"
TOOLCHAIN_MINIMUM_VERSION_gcc="6.0"
TOOLCHAIN_MINIMUM_VERSION_microsoft="19.28.0.0" # VS2019 16.8, aka MSVC 14.28
TOOLCHAIN_MINIMUM_VERSION_xlc="16.1.0.0011"
```

**Comment by ThanHenderson** [comment]
Conversations internally about AIX is that the `xlC` support is for v13 and up (13, 16, & 17), and V2R3 for z/OS.

**Comment by ThanHenderson** [comment]
@pshipton Are `gcc 10.3` for Linux and `VS 2022` for Windows the appropriate versions here?
From the 0.43 release notes: [linux](https://github.com/eclipse-openj9/openj9/blob/775a2c2fe2bffcf62a02d3c70c453d3331686019/doc/release-notes/0.43/0.43.md?plain=1#L54) [windows](https://github.com/eclipse-openj9/openj9/blob/775a2c2fe2bffcf62a02d3c70c453d3331686019/doc/release-notes/0.43/0.43.md?plain=1#L61)

Do you know of any consumers of OMR that are relying on older versions?

**Comment by pshipton** [comment]
The z/OS compile machines for IBM Java 8 are 2.1
Agreed AIX is min xlC v13. We also compile on PPC BE which is also v13. It's being retired hopefully but it's not retired yet.
PPC LE just uses gcc now, not xlC.
Mac is missing from the list, the OpenJ9 compilers are listed in https://eclipse.dev/openj9/docs/openj9_support/
The minimum gcc used for OpenJ9 is 10.3 and the minimum VS used is 2022. That doesn't mean older versions won't work, but we don't test them. Perhaps older version are fine since xlC 13 limits what can be used.

Where did you get the "excerpt from the OpenJ9 building instructions"?

I don't know about other consumers of OMR.
@0xdaryl @mstoodle may know.

**Comment by ThanHenderson** [comment]
Thanks for the clarification, Peter. It'll help me determine which features we could safely use.

The excerpt was from the [extension repos](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/cbd5117c34619bfb4b5e6cf11815a118df44bc9a/doc/building.md?plain=1#L348), not OpenJ9s instructions.

**Comment by 0xdaryl** [comment]
I agree the document should be kept up to date, and thank you @ThanHenderson for looking at that.

While I am generally in favour of continuously raising the bar to allow more C++ features to be used, I think we need to be a bit more gradual about it (e.g., a jump from gcc 4.4.7 to 10.3 is pretty big in one go so perhaps we find something in between).  Also remember that Eclipse OMR is an independent project from Eclipse OpenJ9, and while OpenJ9 is an important consumer, the OMR project can't base all its decisions on OpenJ9's requirements.

Moving up the compiler level is the kind of thing that should be announced to the community for awareness.  I think the OMR release boundaries are appropriate for these kinds of changes (we have the 0.2.0 release planned for end of June, and every 3 months thereafter).  If you have a compiler level proposal in between 4.4.7 and 10.3 for end of June please share for consideration.

To address your question, at one point a couple of years ago I am aware there was a project exploring the use of OMR that required an older gcc version (because their build infrastructure was unable to move up right away).  That may no longer be a requirement for them and enough time has passed that OMR could make the case to gradually increase its minimum supported compiler level.  Perhaps @mstoodle can weigh in on whether this is still a requirement.

**Comment by mstoodle** [comment]
> at one point a couple of years ago I am aware there was a project exploring the use of OMR that required an older gcc version (because their build infrastructure was unable to move up right away). That may no longer be a requirement for them and enough time has passed that OMR could make the case to gradually increase its minimum supported compiler level. Perhaps @mstoodle can weigh in on whether this is still a requirement.

I know the project you're referring to: it's not a requirement any more. I think it more comes down to consistent availability of features across the full set of platforms we and our stakeholders need to support. Otherwise, I'm happy for the bar to move up.

**Comment by ThanHenderson** [comment]
I think a safe bet for the gcc version is 7.3 since that is the version that released with the spectre mitigation patch. And talking to the IBM SDK release team in India, 7.3 is what the JDK8 release still uses.

From a C++ feature standpoint, gcc was C++11 complete by release 4.8, so any release after that is not C++11 feature limiting for platforms compiling with gcc. It will be the other platforms which dictate the supported features, for which I've obtained the following minimum versions:
- Windows: VS2017 (C++14 compliant)
- AIX: xlc-12.01 (a bit nebulous ([compiler reference](https://www.ibm.com/support/pages/system/files/support/swg/swgdocs.nsf/0/330533a7bf467181852579ad001c960f/$FILE/compiler.pdf)))
- Mac: Xcode 12 (Clang 12 -- C++17 compliant)

**Comment by pshipton** [comment]
> And talking to the IBM SDK release team in India, 7.3 is what the JDK8 release still uses.

Just FYI, the OpenJ9 part of the IBM SDK uses 11.2, it's only other components that may still be using 7.3.

**Comment by ThanHenderson** [comment]
I propose bumping the documentation and development to:
- gcc 7.3
- xlC 13.1.3
- Visual Studio 2022
- Xcode 12 (Clang 12)

**Comment by ThanHenderson** [comment]
> Moving up the compiler level is the kind of thing that should be announced to the community for awareness. I think the OMR release boundaries are appropriate for these kinds of changes

@0xdaryl Could we target the 0.3.0 release to update the community/release docs to specify the versions from https://github.com/eclipse/omr/issues/7359#issuecomment-2163534092?
